src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV2.cs

```diff
@@ -289,20 +301,32 @@ private async Task<int> GetPipelineIdAsync(AgentTaskPluginExecutionContext conte
             }
             var definitions = new List<int>() { definition };
 
-            List<Build> list;
-            if (pipelineVersionToDownload == pipelineVersionToDownloadLatest)
-            {
-                list = await buildHttpClient.GetBuildsAsync(project, definitions, tagFilters: tagFilters, queryOrder: BuildQueryOrder.FinishTimeDescending, resultFilter: resultFilter);
-            }
-            else if (pipelineVersionToDownload == pipelineVersionToDownloadLatestFromBranch)
-            {
-                list = await buildHttpClient.GetBuildsAsync(project, definitions, branchName: branchName, tagFilters: tagFilters, queryOrder: BuildQueryOrder.FinishTimeDescending, resultFilter: resultFilter);
-            }
-            else
+            List<Build> list = await buildHttpClient.GetBuildsAsync(project, definitions, branchName: branchName, tagFilters: tagFilters, queryOrder: BuildQueryOrder.FinishTimeDescending, resultFilter: resultFilter);
+
+            if (pipelineVersionToDownload == pipelineVersionToDownloadPullRequestMergeTarget)
             {
-                throw new InvalidOperationException("Unreachable code!");
-            }
+                if (!Int32.TryParse(context.Variables.GetValueOrDefault("system.pullrequest.pullrequestid").Value, out int pullreuestId))
+                {
+                    throw new InvalidOperationException($"Expected system.pullrequest.pullrequestid to be set.");
+                }
+                GitHttpClient gitHttpClient = connection.GetClient<GitHttpClient>();
+                var pullRequest = await gitHttpClient.GetPullRequestByIdAsync(project, pullreuestId);
+                var targetCommitId = pullRequest.LastMergeTargetCommit.CommitId;
 
+                var build = list.FirstOrDefault(build => build.SourceVersion == targetCommitId);
+                if (build == null)
+                {
+                    // No completed builds match. Try searching in-progress builds in case the artifact was published but the build hasn't finished
+                    build = (await buildHttpClient.GetBuildsAsync(project, definitions, branchName: branchName, tagFilters: tagFilters, queryOrder: BuildQueryOrder.FinishTimeDescending, statusFilter: BuildStatus.InProgress))
```

Comment:
We may want to consider expanding the query page size, or fetching next page, on both this and the above query for PRs. It's _extremely_ unlikely the build we're looking for won't be in the most recently few iterations as the PR branch is merged with the latest commit from the target branch as the build starts, but there is an edge case here if there was an extremely high volume of commits into the target branch.

I was initially hoping to set a date filter based on the `targetCommit.Push.Date`, but that seems to be null for cases I tested.